### PR TITLE
samples: nrf9160: https_client: Option to use TF-M and MbedTLS

### PR DESCRIPTION
**Pull request by jarlamsa**
Adding an option to use TF-M and MbedTLS in the https-sample.
Zephyr MbedTLS is used instead of the TLS stack offloaded
in the modem. TF-M is used for keeping the crypto operations
running on the secure side of the device.

Signed-off-by: Jarno Lamsa <jarno.lamsa@nordicsemi.no>

### CODE REVIEW

File: doc/nrf/releases/release-notes-changelog.rst [hunk 1]
```diff
@@ -105,6 +105,10 @@ nRF9160
 
     * Added API to enable modem domain events.
 
+  * :ref:`https_client` sample:
+
+    * Added a possibility to use TF-M and Zephyr MbedTLS instead of using the offloaded TLS stack in modem.
```
Comment (peknis) [inline doc/nrf/releases/release-notes-changelog.rst:110]:
```suggestion
    * Added a possibility to use TF-M and Zephyr Mbed TLS instead of using the offloaded TLS stack in modem.
```

File: doc/nrf/ug_tfm.rst [hunk 2]
```diff
@@ -28,6 +28,7 @@ The TF-M implementation in |NCS| is currently demonstrated in the following samp
 - The :ref:`tfm_hello_world` sample
 - All :ref:`cryptography samples <crypto_samples>` in this SDK
 - A series of :ref:`TF-M integration samples <zephyr:tfm_integration-samples>` available in Zephyr
+- The :ref:`https_client` sample for NRF9160 in this SDK
```
Comment (peknis) [inline doc/nrf/ug_tfm.rst:31]:
```suggestion
- The :ref:`https_client` sample for nRF9160 in this SDK
```

File: samples/nrf9160/https_client/README.rst [hunk 3]
```diff
@@ -40,6 +40,16 @@ This certificate is provided in the :file:`samples/nrf9160/https_client/cert` fo
 To connect to other servers, you might need to provision a different certificate.
 See :ref:`cert_dwload` for more information.
 
+Using MbedTLS and TF-M
+**********************
```
Comment (peknis) [inline samples/nrf9160/https_client/README.rst:44]:
```suggestion
Using Mbed TLS and TF-M
***********************
```

File: samples/nrf9160/https_client/README.rst [hunk 4]
```diff
@@ -40,6 +40,16 @@ This certificate is provided in the :file:`samples/nrf9160/https_client/cert` fo
...
+
+This sample supports using MbedTLS and Trusted Firmware-M (TF-M).
```
Comment (peknis) [inline samples/nrf9160/https_client/README.rst:46]:
```suggestion
This sample supports using Mbed TLS and Trusted Firmware-M (TF-M).
```

File: samples/nrf9160/https_client/README.rst [hunk 5]
```diff
@@ -40,6 +40,16 @@ This certificate is provided in the :file:`samples/nrf9160/https_client/cert` fo
...
+Instead of offloading the TLS sockets in to the modem, the MbedTLS library from Zephyr can be utilized.
```
Comment (peknis) [inline samples/nrf9160/https_client/README.rst:47]:
```suggestion
Instead of offloading the TLS sockets into the modem, you can use the Mbed TLS library from Zephyr.
```

File: samples/nrf9160/https_client/README.rst [hunk 6]
```diff
@@ -40,6 +40,16 @@ This certificate is provided in the :file:`samples/nrf9160/https_client/cert` fo
...
+Using the Zephyr MbedTLS, the offloaded sockets utilizing the modem can still be used. The MbedTLS offers more
+configuration options than is possible when using the offloaded TLS handling.
```
Comment (peknis) [inline samples/nrf9160/https_client/README.rst:49]:
```suggestion
Using the Zephyr Mbed TLS, you can still use the offloaded sockets.
Mbed TLS offers more configuration options than using the offloaded TLS handling.
```

File: samples/nrf9160/https_client/README.rst [hunk 7]
```diff
@@ -40,6 +40,16 @@ This certificate is provided in the :file:`samples/nrf9160/https_client/cert` fo
...
+
+This sample uses the TF-M while using the MbedTLS, thus all the crypto operations are run on the secure side of the device.
```
Comment (peknis) [inline samples/nrf9160/https_client/README.rst:51]:
```suggestion
This sample uses TF-M and Mbed TLS.
Thus, all the crypto operations are run on the secure side of the device.
```

File: samples/nrf9160/https_client/README.rst [hunk 8]
```diff
@@ -48,6 +58,7 @@ Building and running
 
 .. include:: /includes/build_and_run_nrf9160.txt
 
+To build the sample with MbedTLS and TF-M, add following to your west build-command: `-DOVERLAY_CONFIG=overlay-tfm_mbedtls.conf`
```
Comment (peknis) [inline samples/nrf9160/https_client/README.rst:61]:
```suggestion
To build the sample with Mbed TLS and TF-M, add the following to your west build command:

.. code-block:: none

   -DOVERLAY_CONFIG=overlay-tfm_mbedtls.conf
```

File: samples/nrf9160/https_client/README.rst [hunk 9]
```diff
@@ -93,3 +104,7 @@ It uses the following `sdk-nrfxlib`_ library:
 In addition, it uses the following sample:
 
 * :ref:`secure_partition_manager`
+
+This sample offers also possibility to use the TF-M module that can be found in the following location in the |NCS| folder structure:
+
+* ``modules/tee/tfm/``
```
Comment (peknis) [inline samples/nrf9160/https_client/README.rst:110]:
```suggestion
This sample offers also a possibility to use the TF-M module that is at :file:`modules/tee/tfm/` in the the |NCS| folder structure.
```

**Review by peknis — CHANGES_REQUESTED** [review]

File: samples/nrf9160/https_client/README.rst (hunk 7 above)
Comment (jarlamsa) [reply samples/nrf9160/https_client/README.rst:51]: In this section the TF-M is the prerequisite to run the crypto on the secure side, otherwise the crypto functionality of the MbedTLS would be used and ran on the unsecure side. Needs maybe a bit clarification here.

Comment (jarlamsa) [reply samples/nrf9160/https_client/README.rst:51]: Maybe this would be better
"While using TF-M and Mbed TLS using PSA crypto, all the crypto operations are run on the secure side on the device."

Comment (peknis) [reply samples/nrf9160/https_client/README.rst:51]: Using x using y sounds a bit weird. How about:
When using TF-M and Mbed TLS with PSA crypto, ....?

**Comment by jarlamsa** [comment]
Thank you @peknis for your review. I think I have addressed your comments and made changes accordingly. Please re-review.

**Review by peknis — APPROVED** [review]

File: samples/nrf9160/https_client/Kconfig [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+# Copyright (c) 2021 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+menu "HTTPS client sample"
+
+config SAMPLE_TFM_MBEDTLS
+	bool "Use TF-M and MbedTLS"
+	default n
```
Comment (rlubos) [inline samples/nrf9160/https_client/Kconfig:11]: No need for `default n`, please remove.

File: samples/nrf9160/https_client/overlay-tfm_mbedtls.conf [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+CONFIG_MBEDTLS_HEAP_SIZE=32768
+CONFIG_MBEDTLS_TLS_LIBRARY=y
+CONFIG_MBEDTLS_X509_LIBRARY=y
+CONFIG_MBEDTLS_PSA_CRYPTO_C=y
+CONFIG_MBEDTLS_LIBRARY_NRF_SECURITY=y
+CONFIG_NRF_SECURITY_ADVANCED=y
+CONFIG_NORDIC_SECURITY_BACKEND=y
+
+# Optimizations
+CONFIG_SPEED_OPTIMIZATIONS=y
```
Comment (rlubos) [inline samples/nrf9160/https_client/overlay-tfm_mbedtls.conf:23]: Why is this needed? I. e. why don't we use the default, size optimization?

File: samples/nrf9160/https_client/src/main.c [hunk 12]
```diff
@@ -114,6 +114,14 @@ int tls_setup(int fd)
 		TLS_SEC_TAG,
 	};
 
+#if defined(CONFIG_SAMPLE_TFM_MBEDTLS)
+	err = tls_credential_add(tls_sec_tag[0], TLS_CREDENTIAL_CA_CERTIFICATE, cert, sizeof(cert));
+
```
Comment (rlubos) [inline samples/nrf9160/https_client/src/main.c:119]: No need for this empty line.

**Review by rlubos — COMMENTED** [review]
Just some nits

**Comment by jarlamsa** [comment]
Thank you @rlubos for your review. I think I have addressed all your comments, please re-review.

**Review by rlubos — APPROVED** [review]

File: doc/nrf/releases/release-notes-changelog.rst [hunk 13]
```diff
@@ -105,6 +105,10 @@ nRF9160
 
     * Added API to enable modem domain events.
 
+  * :ref:`https_client` sample:
+
+    * Added a possibility to use TF-M and Zephyr Mbed TLS instead of using the offloaded TLS stack in modem.
+
```
Comment (lemrey) [inline doc/nrf/releases/release-notes-changelog.rst:111]: If I remember correctly we did not track changes to samples in the changelog?

File: samples/nrf9160/https_client/src/main.c [hunk 14]
```diff
@@ -114,6 +114,13 @@ int tls_setup(int fd)
 		TLS_SEC_TAG,
 	};
 
+#if defined(CONFIG_SAMPLE_TFM_MBEDTLS)
+	err = tls_credential_add(tls_sec_tag[0], TLS_CREDENTIAL_CA_CERTIFICATE, cert, sizeof(cert));
+	if (err) {
+		return err;
+	}
+#endif
```
Comment (lemrey) [inline samples/nrf9160/https_client/src/main.c:122]: Why not move this next to code that provisions the certificates to the modem?

File: doc/nrf/releases/release-notes-changelog.rst (hunk 13 above)
Comment (rlubos) [reply doc/nrf/releases/release-notes-changelog.rst:111]: Why not? It's a significant change to the sample, and we do have changelog entries for samples in the past release notes.

**Review by SeppoTakalo — APPROVED** [review]

File: samples/nrf9160/https_client/src/main.c (hunk 14 above)
Comment (jarlamsa) [reply samples/nrf9160/https_client/src/main.c:122]: Mainly because I didn't want to introduce the `tls_sec_tag` again or make it global.

Comment (lemrey) [reply samples/nrf9160/https_client/src/main.c:122]: Sorry, I meant why don't we put this piece of code next to the bit that calls `cert_provision()` a few lines below?
This way we show the two different calls that provision the certificates one next to the other.

File: doc/nrf/releases/release-notes-changelog.rst (hunk 13 above)
Comment (lemrey) [reply doc/nrf/releases/release-notes-changelog.rst:111]: Ok, I remember wrongly then.

File: samples/nrf9160/https_client/src/main.c (hunk 14 above)
Comment (jarlamsa) [reply samples/nrf9160/https_client/src/main.c:122]: The call for `cert_provision()` is on the main-function and this call is in the `tls_setup()` -function. The `tls_sec_tag` is introduced in the `tls_setup()`-function as well. I know what you mean, but I wanted to make only minimal changes to the sample.

**Comment by jarlamsa** [comment]
Rebased and resolved conflicting changelog

**Review by lemrey — APPROVED** [review]
